contracts/extensions/OneTxPayment.sol

```diff
@@ -206,6 +212,88 @@ contract OneTxPayment is ColonyExtension {
     }
   }
 
+  function calculateUniqueAmounts(
+    address[] memory _tokens,
+    uint256[] memory _amounts
+  )
+    internal
+    pure
+    returns (uint256, address[] memory, uint256[] memory)
+  {
+    uint256 uniqueTokensIdx;
+    address[] memory uniqueTokens = new address[](MAX_TOKENS);
+    uint256[] memory uniqueAmounts = new uint256[](MAX_TOKENS);
+
+    for (uint256 i; i < _tokens.length; i++) {
+      bool isMatch;
+      for (uint256 j; j < uniqueTokensIdx && !isMatch; j++) {
```

Comment:
I thought about that, but I don't fully agree since we are iterating over the `uniqueTokensIndex` and running over that index is also a legitimate condition (either we max out the index OR find a match). So if we do a `while` we'll have to include that `for` functionality anyway, since this is really a hybrid situation -- either we make it a `for` and toss in `while`-style functionality, or make it a `while` and toss in `for`-style functionality.

Reply:
I think `while-that-has-some-for-functionality` is clearer than `for-that-has-while-functionality` though - here the additional functionality is buried in the middle of a line, which could be overlooked (I nearly did).

I hope the compiler boils it all down to the same thing though :smile: